### ISSUE TITLE
[Discussion] Consider removing `--mount type=devpts,destination=/dev/pts`

### ISSUE THREAD

**Issue by alexpdp7** [89luca89/distrobox#688]
When using distrobox with Podman nested inside an LXC container, `distrobox enter` does not work, although in general `podman` works.

If I patch distrobox not to use `--mount type=devpts,destination=/dev/pts`, then it seems to work OK.

I traced the introduction of this option to toolbx, which introduced this in https://github.com/containers/toolbox/issues/568#issuecomment-708595672 . However, it seems that this change does not address the underlying problem- the issue is still open.

Given that distrobox supports Docker, which does not even support this option... perhaps we could remove it? It doesn't seem to serve its original purpose, and it causes issues in some (admittedly "unsupported"- and I think it's fine not to support nesting in LXC) environments.

(I suspect there might be an underlying "bug" in podman/crun that's triggered by using this flag, but it's just a hunch.)

**Comment by alexpdp7** [comment]
This seems to be related to https://github.com/89luca89/distrobox/pull/671 ? The actual error I get is this https://github.com/alexpdp7/alexpdp7/issues/10#issuecomment-1493430815 , which looks somewhat similar.

There seems to be something interesting lurking around, but I don't know enough to pin it down :(

**Comment by ericcurtin** [comment]
Hit the same issue, trying to run CentOS Stream 9 QT-based UI apps using distrobox + podman + crun on debian bookworm lxc environment on a aarch64 Chromebook running Chrome OS (hope that makes sense), but I ran into many issues to be honest, this is just one.

**Comment by 89luca89** [comment]
As of commit f1913c2 the init is now doing the devpts internally
can anyone test if this still creates problems for LXC?

**Comment by ericcurtin** [comment]
Different bug now, on latest version of git main branch:

```
$ distrobox enter -r fedora-38
Container fedora-38 is not running.
Starting container fedora-38
run this command to follow along:

 sudo podman logs -f fedora-38

 Error: could not start entrypoint.
You must run  entrypoint inside a container!
distrobox-init should only be used as an entrypoint for a distrobox!

This is not intended to be used manually, but instead used by distrobox-enter
to set up the container's entrypoint.
Error: An error occurred
```

**Comment by ericcurtin** [comment]
This would be a significant feature for ChromeOS if we could fix this, it would mean you could run almost any Linux distro's UI app on ChromeOS

**Comment by 89luca89** [comment]
How is your setup?

Lxc with inside rootless podman?

**Comment by ericcurtin** [comment]
Yeah... That's what ChromeOS gives you out of the box (well you can choose to install either docker or podman)... You are given an LXC Debian environment that can run UI apps with sound, etc.

But there's no easy way to switch from Debian to other distros in this environment, that's why distrobox is super handy here.

I've gotten distrobox working with non-UI apps on it very well, by more of less deleting offending lines if memory serves me right on how I hacked it.

But UI apps don't work there at present.

**Comment by ericcurtin** [comment]
And just to confirm UI apps work fine of course outside the distrobox environment, in the Debian environment provided by ChromeOS

**Comment by 89luca89** [comment]
Not sure what is different in Chromeos, but on ubuntu, using LXC (via lxd) + Distrobox with nested rootless podman seems to work as intended:

![image](https://github.com/89luca89/distrobox/assets/598882/79387e32-9ce3-4455-b8bc-71cff26068be)

**Comment by ericcurtin** [comment]
Maybe we could close this issue if it solves @alexpdp7 's case and open a separate issue called "ChromeOS support" or something like that?

**Comment by 89luca89** [comment]
We could yes
But anyway it seems undoable for me to debug or support something that I don't own nor I can easily reproduce (in VM or such)

**Comment by alexpdp7** [comment]
Hmmm,

I just did:

```
[alex@ws ~]$ curl -s https://raw.githubusercontent.com/89luca89/distrobox/main/install | sudo sh -s -- --next
[alex@ws ~]$ distrobox create
[alex@ws ~]$ distrobox enter my-distrobox
Container my-distrobox is not running.
Starting container my-distrobox
run this command to follow along:

 podman logs -f my-distrobox

 Error: could not start entrypoint.
You must run  entrypoint inside a container!
distrobox-init should only be used as an entrypoint for a distrobox!

This is not intended to be used manually, but instead used by distrobox-enter
to set up the container's entrypoint.
Error: An error occurred
```

This is in a Rocky Linux 9 LXC Container in Proxmox. I'll do more digging.

**Comment by alexpdp7** [comment]
OK, `distrobox create` starts a container, but dies. I ran `podman inspect` on the container to extract the command used to create the container:

```
'podman create --hostname my-distrobox.ws.h1.int.pdp7.net --name my-distrobox --privileged --security-opt label=disable --user root:root --ipc host --network host --pid host --label manager=distrobox --env SHELL=bash --env HOME=/home/alex --mount type=tmpfs,destination=/run --volume /:/run/host:rslave --volume /dev:/dev:rslave --volume /sys:/sys:rslave --volume /tmp:/tmp:rslave --volume /usr/local/bin/distrobox-init:/usr/bin/entrypoint:ro --volume /usr/local/bin/distrobox-export:/usr/bin/distrobox-export:ro --volume /usr/local/bin/distrobox-host-exec:/usr/bin/distrobox-host-exec:ro --volume /home/alex:/home/alex:rslave --volume /dev/pts --volume /dev/null:/dev/ptmx --volume /var/log/journal --volume /run/user/1284000001:/run/user/1284000001:rslave --volume /etc/hosts:/etc/hosts:ro --volume /etc/resolv.conf:/etc/resolv.conf:ro --ulimit host --annotation run.oci.keep_original_groups=1 --userns keep-id --entrypoint /usr/bin/entrypoint registry.fedoraproject.org/fedora-toolbox:38 --verbose --name alex --user 1284000001 --group 1284000001 --home /home/alex --init 0 --nvidia 0 --pre-init-hooks  --additional-packages  --'
```

I tried:

```
[alex@ws ~]$ podman run -it --rm --hostname my-distrobox.ws.h1.int.pdp7.net --name my-distrobox --privileged --security-opt label=disable --user root:root --ipc host --network host --pid host --label manager=distrobox --env SHELL=bash --env HOME=/home/alex --mount type=tmpfs,destination=/run --volume /:/run/host:rslave --volume /dev:/dev:rslave --volume /sys:/sys:rslave --volume /tmp:/tmp:rslave --volume /usr/local/bin/distrobox-init:/usr/bin/entrypoint:ro --volume /usr/local/bin/distrobox-export:/usr/bin/distrobox-export:ro --volume /usr/local/bin/distrobox-host-exec:/usr/bin/distrobox-host-exec:ro --volume /home/alex:/home/alex:rslave --volume /dev/pts --volume /dev/null:/dev/ptmx --volume /var/log/journal --volume /run/user/1284000001:/run/user/1284000001:rslave --volume /etc/hosts:/etc/hosts:ro --volume /etc/resolv.conf:/etc/resolv.conf:ro --ulimit host --annotation run.oci.keep_original_groups=1 --userns keep-id --entrypoint /usr/bin/entrypoint registry.fedoraproject.org/fedora-toolbox:38 --verbose --name alex --user 1284000001 --group 1284000001 --home /home/alex --init 0 --nvidia 0 --pre-init-hooks  --additional-packages  --
Error: OCI runtime error: crun: unlockpt: Invalid argument
```

If I remove `--volume /dev/null:/dev/ptmx`, then:

```
[alex@ws ~]$ podman run -it --rm --hostname my-distrobox.ws.h1.int.pdp7.net --name my-distrobox --privileged --security-opt label=disable --user root:root --ipc host --network host --pid host --label manager=distrobox --env SHELL=bash --env HOME=/home/alex --mount type=tmpfs,destination=/run --volume /:/run/host:rslave --volume /dev:/dev:rslave --volume /sys:/sys:rslave --volume /tmp:/tmp:rslave --volume /usr/local/bin/distrobox-init:/usr/bin/entrypoint:ro --volume /usr/local/bin/distrobox-export:/usr/bin/distrobox-export:ro --volume /usr/local/bin/distrobox-host-exec:/usr/bin/distrobox-host-exec:ro --volume /home/alex:/home/alex:rslave --volume /dev/pts --volume /var/log/journal --volume /run/user/1284000001:/run/user/1284000001:rslave --volume /etc/hosts:/etc/hosts:ro --volume /etc/resolv.conf:/etc/resolv.conf:ro --ulimit host --annotation run.oci.keep_original_groups=1 --userns keep-id --entrypoint /usr/bin/entrypoint registry.fedoraproject.org/fedora-toolbox:38 --verbose --name alex --user 1284000001 --group 1284000001 --home /home/alex --init 0 --nvidia 0 --pre-init-hooks  --additional-packages  --
Error: crun: open /dev/pts/4: No such file or directory: OCI runtime attempted to invoke a command that was not found
```

If I remove `--volume /dev/pts` on top, then:

```
[alex@ws ~]$ podman run -it --rm --hostname my-distrobox.ws.h1.int.pdp7.net --name my-distrobox --privileged --security-opt label=disable --user root:root --ipc host --network host --pid host --label manager=distrobox --env SHELL=bash --env HOME=/home/alex --mount type=tmpfs,destination=/run --volume /:/run/host:rslave --volume /dev:/dev:rslave --volume /sys:/sys:rslave --volume /tmp:/tmp:rslave --volume /usr/local/bin/distrobox-init:/usr/bin/entrypoint:ro --volume /usr/local/bin/distrobox-export:/usr/bin/distrobox-export:ro --volume /usr/local/bin/distrobox-host-exec:/usr/bin/distrobox-host-exec:ro --volume /home/alex:/home/alex:rslave --volume /var/log/journal --volume /run/user/1284000001:/run/user/1284000001:rslave --volume /etc/hosts:/etc/hosts:ro --volume /etc/resolv.conf:/etc/resolv.conf:ro --ulimit host --annotation run.oci.keep_original_groups=1 --userns keep-id --entrypoint /usr/bin/entrypoint registry.fedoraproject.org/fedora-toolbox:38 --verbose --name alex --user 1284000001 --group 1284000001 --home /home/alex --init 0 --nvidia 0 --pre-init-hooks  --additional-packages  --
You must run  entrypoint inside a container!
distrobox-init should only be used as an entrypoint for a distrobox!

This is not intended to be used manually, but instead used by distrobox-enter
to set up the container's entrypoint.
Error: An error occurred
```

, which means that the container works, but the entrypoint doesn't...

```
[alex@ws ~]$ podman run -it --rm --hostname my-distrobox.ws.h1.int.pdp7.net --name my-distrobox --privileged --security-opt label=disable --user root:root --ipc host --network host --pid host --label manager=distrobox --env SHELL=bash --env HOME=/home/alex --mount type=tmpfs,destination=/run --volume /:/run/host:rslave --volume /dev:/dev:rslave --volume /sys:/sys:rslave --volume /tmp:/tmp:rslave --volume /usr/local/bin/distrobox-init:/usr/bin/entrypoint:ro --volume /usr/local/bin/distrobox-export:/usr/bin/distrobox-export:ro --volume /usr/local/bin/distrobox-host-exec:/usr/bin/distrobox-host-exec:ro --volume /home/alex:/home/alex:rslave --volume /var/log/journal --volume /run/user/1284000001:/run/user/1284000001:rslave --volume /etc/hosts:/etc/hosts:ro --volume /etc/resolv.conf:/etc/resolv.conf:ro --ulimit host --annotation run.oci.keep_original_groups=1 --userns keep-id registry.fedoraproject.org/fedora-toolbox:38
[root@my-distrobox /]# /usr/bin/entrypoint
You must run  entrypoint inside a container!
distrobox-init should only be used as an entrypoint for a distrobox!

This is not intended to be used manually, but instead used by distrobox-enter
to set up the container's entrypoint.
Error: An error occurred
```

What I don't understand is the probing for running inside a container. It seems to check for `/run/.containerenv` and `/.dockerenv`. But I suspect `podman` does not create `/.dockerenv` and `/run` is supposed to be empty because of `--mount type=tmpfs,destination=/run`?

**Comment by ericcurtin** [comment]
I now have working UI apps on distrobox on ChromeOS, cool 😄 

Fully working now.

**Comment by ericcurtin** [comment]
After #916 it works

**Comment by 89luca89** [comment]
Great stuff thanks everyone for checking and testing!

Set the @ericcurtin PR as solver for this issue :+1:
